File: 2021/01/ruby/solution.rb
# Advent of Code 2021
# Day 1: Sonar Sweep

module Day01
  module_function

  def run
    puts "Part 1: #{part_one}"
    puts "Part 2: #{part_two}"
  end

  def part_one
    count_increases(depths)
  end

  def part_two
    count_increases(sliding_sums(depths))
  end

  def count_increases(numbers)
    numbers.lazy
      .each_cons(2)
      .select { |x, y| x < y }
      .count
  end

  def sliding_sums(numbers)
    numbers.lazy
      .each_cons(3)
      .collect(&:sum)
  end

  def depths
    input.each_line.map(&:to_i)
  end

  def input
    File.open("../input.txt")
  end
end

if $PROGRAM_NAME == __FILE__
  Day01.run
end

File: 2021/01/swift/solution.swift
import Foundation

let input = try! String(contentsOfFile: "../input.txt")
let depths = input.split(separator: "\n").map { Int($0)! }

var incs = 0
var sums = 0
for curr in 1..<depths.count {
    if depths[curr-1] < depths[curr] {
        incs += 1
    }

    if curr >= 3 {
        let a = depths[curr-3] + depths[curr-2] + depths[curr-1]
        let b = depths[curr-2] + depths[curr-1] + depths[curr]
        if b > a {
            sums += 1
        }
    }
}

print("Part 1: \(incs)")
print("Part 2: \(sums)")
